shell/platform/darwin/ios/platform_service_ios.mm

```diff
@@ -0,0 +1,105 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/darwin/ios/platform_service_ios.h"
+
+#include "base/mac/scoped_nsautorelease_pool.h"
+#include "lib/ftl/logging.h"
+
+#import <AudioToolbox/AudioToolbox.h>
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
+namespace shell {
+
+namespace {
+
+void Vibrate(NSMutableDictionary* result) {
+  AudioServicesPlayAlertSound(kSystemSoundID_Vibrate);
+}
+
+void LaunchURL(NSString* urlString, NSMutableDictionary* result) {
+  NSURL* url = [NSURL URLWithString:urlString];
+
+  UIApplication* application = [UIApplication sharedApplication];
+
+  bool launched = [application canOpenURL:url] && [application openURL:url];
+
+  [result setObject:(launched ? @"true" : @"false") forKey:@"launched"];
+}
+
+void GetDirectoryOfType(NSString* pathType, NSMutableDictionary* result) {
+  NSSearchPathDirectory dir;
+  if ([pathType isEqual:@"temporary"]) {
+    dir = NSCachesDirectory;
+  } else if ([pathType isEqual:@"documents"]) {
+    dir = NSDocumentDirectory;
+  } else {
+    FTL_LOG(ERROR) << "Unsupported path type " << [pathType UTF8String];
+    return;
+  }
+  NSArray* paths =
+      NSSearchPathForDirectoriesInDomains(dir, NSUserDomainMask, YES);
+
+  if (paths.count == 0) {
+    return;
+  }
+
+  [result setObject:paths.firstObject forKey:@"path"];
+}
+
+}  // namespace
+
+PlatformServiceIOS::PlatformServiceIOS() = default;
+PlatformServiceIOS::~PlatformServiceIOS() = default;
+
+void PlatformServiceIOS::Process(std::string data,
+                                 std::function<void(std::string)> callback) {
+  base::mac::ScopedNSAutoreleasePool pool;
+
+  NSError* err = nil;
+  NSString* dataAsNSString = [NSString stringWithUTF8String:data.c_str()];
```

Comment:
@chinmaygarde - how do I go from std::string containing UTF-8 encoded data to NSData (for NSJSONSerialization) in a way that doesn't involve copies? This seems to work but I'm pretty sure it's copying the data at least once, which seems dumb